File: src/interpolate.py
#!/usr/bin/env python3
# coding: utf-8
import numpy as np
from sys import argv
from os.path import basename, splitext
import glob
import pandas as pd

def anynan(x, axis=0):
    return np.any(np.isnan(x), axis=axis)

def nNaN(curve):
    """
    Get number of points in a curve that contain NaN.
    """
    return np.sum(anynan(curve, axis=1))

def interpolate_nan(curve):
    """
    Replace points containing NaN with linear interpolation of neighbors.
    curve: matrix with xyz
    """
    if anynan(curve[0]) or anynan(curve[-1]):
        raise NotImplementedError
    
    curve_interp = []
    i = 0
    while i < len(curve):
        if not anynan(curve[i]):
            curve_interp.append(curve[i])
            i += 1
        else:
            # n == number of nans in a row to convert.
            n = 1
            while anynan(curve[i+n]): n += 1
            for j in range(1, n+1):
                curve_interp.append((1-j/(n+1))*curve[i-1] + (j/(n+1))*curve[i+n])
            i += n
    
    return np.asarray(curve_interp)

def interpolate(curve, dist=0.02):
    """
    Add interpolated points between xyzs of curve.
    dist: place points approximately this far apart
    """
    curve_interp = []
    interp_idx = []
    
    for i in range(len(curve)-1):
        # add first real point of a segment
        curve_interp.append(curve[i])
        interp_idx.append(False)
        
        d = np.linalg.norm(curve[i] - curve[i+1])
        # number of interpolation points propertional to distance between two real points.
        n = int(np.ceil(d / dist))
        # starting from 1 means we don't add the first real point twice
        for j in range(1, n):
            curve_interp.append((1-j/n)*curve[i] + (j/n)*curve[i+1])
            interp_idx.append(True)
        
    # add terminal real point
    curve_interp.append(curve[-1])
    interp_idx.append(False)
    
    return np.asarray(curve_interp), np.asarray(interp_idx)

def main():   
    # USAGE: ./interpolate.py INDIR/ OUTDIR/ DIST MAXNAN
    # DIST: approx distance between interpolated points
    # MAXNAN: number of points with NaNs allowed.
    # Outfiles will be given the same name as infiles but placed in OUTDIR
    INDIR, OUTDIR, DIST, MAXNAN = argv[1], argv[2], float(argv[3]), int(argv[4])

    for fname in glob.glob(f"{INDIR}/*.npy"):
        curve = np.load(fname)
        name = splitext(basename(fname))[0]
        
        if nNaN(curve) > 0:
            if nNaN(curve) > MAXNAN:
                print(f"{nNaN(curve)} > {MAXNAN} NaNs. SKIP {fname}")
                continue
            else:
                print(f"{nNaN(curve)} <= {MAXNAN} NaNs. Allowing {fname}")
        
            try: curve = interpolate_nan(curve)
            except NotImplementedError:
                print("Not implemented first or last NaN point.")
                continue
        
        curve_interp, interp_idx = interpolate(curve, DIST)
        assert nNaN(curve_interp) == 0, nNaN(curve_interp)
        
        df = pd.DataFrame(curve_interp, columns=list("xyz"))
        df["interp"] = interp_idx.astype(int)
        df.to_csv(f"{OUTDIR}/{name}.tsv", sep='\t', index=False)
        print(f"{name}:\t{len(curve)}\t->\t{len(curve_interp)}")

if __name__ == "__main__":
    main()

File: src/interpolate.jl
#!/usr/bin/env julia
using ArgParse
using DelimitedFiles
using NPZ
include("glob.jl") # glob that accepts abspath

parser = ArgParseSettings(description="""
Take xyz curve and interpolate linearly between points, given a fixed number of 
required output points.
""")

@add_arg_table parser begin
    "infiles"
    arg_type = String
    nargs = '+'
    help = "Infiles, indir, or glob pattern for infiles. 
    .npy, .csv, or .tsv without headers."
    "--out", "-o"
    arg_type = String
    help = "Output directory. 
    Files are written with the same names as infiles except extension will be .tsv. 
    If it doesn't exist it will be created."
    "--points", "-n"
    arg_type = Int
    help = "Number of points in each output file."
    "--nans", "-N"
    arg_type = Int
    default = 0
    help = "Max number of allowed NaNs in a curve. Curves with more are skipped. 
    NaNs are replaced by points interpolated between the point before and after.
    Curves with NaN in first or last point are skipped."
end

# if run as script
if abspath(PROGRAM_FILE) == @__FILE__
    args = ARGS
else
    # if interactive, do an example. You may need to unzip pointClouds.zip first.
    ROOT = readchomp(`git root`)
    sArgs = "$ROOT/Data/HCT116_chr21-28-30Mb_untreated/pointClouds -o lars -n 100 -N 1"
    args = split(sArgs, ' ')
end
args = parse_args(args, parser, as_symbols=true)
for (k,v) in args println("# $k = $v") end
args = NamedTuple(args)

# if indir given
infiles = glob(args.infiles)
if isdir(infiles) infiles = glob(infiles[1] .* "/*") end

# a consistent fileformat should be given
extMatches = [endswith.(infiles, ext) for ext in [".npy", ".tsv", ".csv"]] .|> any |> sum
@assert extMatches == 1 "$extMatches extension matches. There should be exactly 1 match (.npy, .tsv or .csv)."

# read
npzs = infiles[endswith.(infiles, ".npy")]
tsvs = infiles[endswith.(infiles, ".tsv")]
csvs = infiles[endswith.(infiles, ".csv")]
names = basename.([npzs; tsvs; csvs])
curves = [npzread.(npzs); readdlm.(tsvs, '\t'); readdlm.(csvs, ',')]

@assert all(size.(curves, 2) .== 3) "All files should have 3 columns."

nIn = maximum(size.(curves, 1))
@assert nIn .<= args.points "-n/--points=$(args.points) should be more than curve lengths=$nIn."

# NaN filters
anynan(v::AbstractArray) = any(isnan.(v))
anynan(v::AbstractArray, dims) = any(isnan.(v); dims=dims)
nnan(curve::AbstractMatrix) = sum(anynan(curve, 2))
hasTerminalNan(curve::AbstractMatrix) = anynan(curve[[1,end], :])

nanFilt = nnan.(curves) .<= args.nans
curves = curves[nanFilt]
names = names[nanFilt]
@info "NaNs > $(args.nans) filter: $(length(nanFilt)) -> $(length(curves))"

terminalNan = hasTerminalNan.(curves)
curves = curves[.!terminalNan]
names = names[.!terminalNan]
@info "terminal NaN filter: $(length(terminalNan)) -> $(length(curves))"

"""
Get point linearly interpolated between point at index "start" and "stop" with 
index "between" where between=start will return point start and between=stop 
will return point stop.
"""
function interp(curve::AbstractMatrix, start::Int, stop::Int, between::Int)
    interp(curve, start, stop, (between - start) / (stop - start))
end
function interp(curve::AbstractMatrix, start::Int, stop::Int, frac::Float64)
    (1-frac) * curve[start,:] + frac * curve[stop,:]
end

"""
Replace points containing NaN with linear interpolation of neighbors.
curve: xyz
"""
function interpNan(curve::Matrix)
    anynan(curve[[1,end], :]) && error("Not implemented.")
    
    n = size(curve, 1)
    
    out = []
    for (i, point) ∈ enumerate(eachrow(curve))
        if !anynan(point)
            push!(out, point)
        else
            # find nearest non-nan neighbors
            start = (1:i-1)[vec(.!anynan(curve[1:i-1, :], 2))] |> last
            stop  = (i+1:n)[vec(.!anynan(curve[i+1:n, :], 2))] |> first
            push!(out, interp(curve, start, stop, i))
        end
    end
    
    hcat(out...)'
end 

if any(anynan.(curves))
    nanRows = anynan.(curves, 2)
    curves = interpNan.(curves)
else
    nanRows = nothing
end

"Euclidean distance between each consecutive point (row) in a curve."
function consecutiveEuclidean(curve::AbstractMatrix)
    sqrt.(sum(diff(curve; dims=1) .^ 2; dims=2))
end

"""
Interpolate between points in a curve of xyzs resulting in n point.
"""
function interp(curve::AbstractMatrix, nOut::Int)
    nIn = size(curve, 1)
    nNew = nOut - nIn
    
    # decide how many points to add between each real point based on the 
    # distances between points
    dists = consecutiveEuclidean(curve)
    nBetween = zeros(Int, nIn - 1)
    for i in 1:nNew
        gaps = dists ./ (nBetween .+ 1) |> vec
        nBetween[argmax(gaps)] += 1
    end
    
    # indicator vector of interpolation
    interpIdx = trues(nOut)
    interpIdx[cumsum(1 .+ [0; nBetween])] .= false
    
    # add the real and interpolated points iteratively
    out = []
    for i in 1:nIn-1
        push!(out, curve[i, :])
        nb = nBetween[i]
        for ii in 1:nb
            push!(out, interp(curve, i, i+1, ii/(nb+1)))
        end
    end
    push!(out, curve[end, :])
    
    hcat(out...)', interpIdx
end

interps = interp.(curves, args.points)
# nanRows are only for uninterped points, update the indices to the 
# interpolated output size.
if nanRows !== nothing
    tmp = nanRows
    nanRows = [falses(args.points) for _ in 1:length(interps)]
    for (i, (_, interpIdx)) in enumerate(interps)
        nanRows[i][.!interpIdx] .= tmp[i]
    end
end

# write
delim, ext = '\t', ".tsv"
header = ["x", "y", "z", "interp"]
nanRows === nothing || push!(header, "nan")
header = join(header, delim) * '\n'

ispath(args.out) || mkdir(args.out)

for i in 1:length(interps)
    interped, interpIdx = interps[i]
    open(args.out * "/" * splitext(names[i])[1] * ext; write=true) do f
        write(f, header)
        mat = nanRows === nothing ? Real[interped interpIdx] : Real[interped interpIdx nanRows[i]]
        writedlm(f, mat)
    end
end

